components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimDAO.java

```diff
@@ -232,4 +233,80 @@ protected void deleteClaimProperties(Connection connection, int claimId, int ten
+        try (PreparedStatement prepStmt = connection.prepareStatement(query)) {
+            prepStmt.setInt(1, claimId);
+            prepStmt.setInt(3, tenantId);
+            
+            for (String propertyName : claimPropertyNames) {
+                prepStmt.setString(2, propertyName);
+                prepStmt.addBatch();
+            }
+            prepStmt.executeBatch();
+        } catch (SQLException e) {
+            throw new ClaimMetadataException("Error while deleting claim properties: " + claimPropertyNames, e);
+        }
+    }
+
+    /**
+     * Updates claim properties atomically by performing deletions and additions within a single transaction.
+     * This method ensures that all operations either complete successfully or are rolled back entirely.
+     *
+     * @param connection              Database connection to be used
+     * @param claimId                 ID of the claim whose properties are being updated
+     * @param newClaimProperties      Map of new claim properties to be added (property name to value mapping)
+     * @param claimPropertiesToDelete Set of claim property names to be deleted
+     * @param tenantId                ID of the tenant
+     * @throws ClaimMetadataException If an error occurs during the transaction or while managing claim properties
+     */
+    protected void updateClaimPropertiesAtomically(Connection connection, int claimId,
+                                                   Map<String, String> newClaimProperties,
+                                                   Set<String> claimPropertiesToDelete,
+                                                   int tenantId) throws ClaimMetadataException {
+
+        boolean autoCommit = false;
+        SQLException resetAutoCommitException = null;
+        
+        try {
+            autoCommit = connection.getAutoCommit();
+            connection.setAutoCommit(false);
+
+            if (claimPropertiesToDelete != null && !claimPropertiesToDelete.isEmpty()) {
+                deleteClaimProperties(connection, claimId, claimPropertiesToDelete, tenantId);
+            }
+
+            if (newClaimProperties != null && !newClaimProperties.isEmpty()) {
+                addClaimProperties(connection, claimId, newClaimProperties, tenantId);
+            }
+
+            IdentityDatabaseUtil.commitTransaction(connection);
+        } catch (SQLException e) {
+            IdentityDatabaseUtil.rollbackTransaction(connection);
+            throw new ClaimMetadataException("Error while updating claim properties atomically", e);
+        } finally {
+            try {
+                connection.setAutoCommit(autoCommit);
+            } catch (SQLException e) {
+                resetAutoCommitException = e;
+            }
+        }
+        
+        if (resetAutoCommitException != null) {
+            throw new ClaimMetadataException("Error occurred while resetting auto-commit state", 
+                    resetAutoCommitException);
+        }
```

Comment:
Can't we get rid of `resetAutoCommitException` variable by refactoring the code as follows?

```suggestion
            } catch (SQLException e) {
                    throw new ClaimMetadataException("Error occurred while resetting auto-commit state", 
                    e);
            }
        }
```

Reply:
Ack, thanks for the explanation.